### PR TITLE
Add pure-retrieval baseline (no sequential frame lookahead) and NMS

### PR DESCRIPTION
**Pull request by johnwlambert**

### CODE REVIEW

File: gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py [hunk 1]
```diff
@@ -15,11 +15,16 @@
...
-    """Keypoint aggregator with de-duplication."""
+    """Keypoint aggregator with de-duplication of keypoints within each image."""
 
-    def __init__(self) -> None:
-        """Initialize global variables"""
+    def __init__(self, nms_radius: float = 3) -> None:
+        """Initialize global variables.
+
+        Args:
+            nms_radius: Radius (in pixels) to use when merging detections within the same view.
```
Comment (travisdriver) [inline gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py:24]: Non-maximum suppression makes it seem like tracks are being suppressed instead of merged. Maybe changed to `merge_radius`?

File: gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py [hunk 2]
```diff
@@ -46,8 +51,8 @@ def append_unique_keypoints(
         for k, uv in enumerate(keypoints.coordinates):
             diff_norms = np.linalg.norm(per_image_kpt_coordinates[i] - uv, axis=1)
             # TODO(johnwlambert,ayushbaid): test loosening threshold below to some epsilon.
-            is_identical = np.any(diff_norms == 0)
-            if len(per_image_kpt_coordinates[i]) > 0 and is_identical:
+            is_duplicate = np.any(diff_norms <= self.nms_radius)
+            if len(per_image_kpt_coordinates[i]) > 0 and is_duplicate:
```
Comment (travisdriver) [inline gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py:55]: Maybe we should use the average coordinate?

File: gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py (hunk 1 above)
Comment (johnwlambert) [reply gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py:24]: Updated to `nms_merge_radius` -- hopefully that captures both.

File: gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py (hunk 2 above)
Comment (johnwlambert) [reply gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py:55]: Left a TODO, thanks.

File: gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py [hunk 3]
```diff
@@ -15,11 +15,17 @@
...
 
-    def __init__(self) -> None:
-        """Initialize global variables"""
+    def __init__(self, nms_merge_radius: float = 3) -> None:
+        """Initialize global variables.
+
+        Args:
+            nms_merge_radius: Radius (in pixels) to use when merging detections within the same view (image).
+                Note that tracks are merged, not suppressed.
+        """
```
Comment (travisdriver) [inline gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py:26]: I really think taking NMS out of the name would reduce any  confusion about supression. I think just `merge_radius` is fine, and then you wouldn't have to put the note.

File: scripts/benchmark_wildcat.sh [hunk 4]
```diff
@@ -99,7 +100,7 @@ for num_matched in ${num_matched_sizes[@]}; do
 					--output_root $OUTPUT_ROOT \
 					--max_resolution 760 \
 					2>&1 | tee $OUTPUT_ROOT/out.log
-				elif [[ $loader == *"olsson"* ]]
+				elif [[ $loader == *"colmap"* ]]
```
Comment (travisdriver) [inline scripts/benchmark_wildcat.sh:103]: What is this change for?

File: gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py (hunk 3 above)
Comment (travisdriver) [reply gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py:26]: Also, have you done any testing with respect to the merge radius? Is 3 pixels the best value you saw?

File: gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py (hunk 1 above)
Comment (travisdriver) [reply gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py]: Non-maximum suppression merge radius seems like a bit of an oxymoron...

File: scripts/benchmark_wildcat.sh (hunk 4 above)
Comment (johnwlambert) [reply scripts/benchmark_wildcat.sh:103]: There was a typo in the elif clause -- the string "olsson" was being used to check if we should run the colmap loader.

File: gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py [hunk 5]
```diff
@@ -15,11 +15,17 @@
...
         self.duplicates_found = 0
+        self.nms_merge_radius = nms_merge_radius
 
     def append_unique_keypoints(
         self, i: int, keypoints: Keypoints, per_image_kpt_coordinates: Dict[Tuple[int, int], np.ndarray]
```
Comment (travisdriver) [inline gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py:31]: I believe the type hint for `per_image_kpt_coordinates` should be `Dict[int, np.ndarray]`

File: gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py [hunk 6]
```diff
@@ -99,10 +105,11 @@ def aggregate(
             per_image_kpt_coordinates, i2_indices = self.append_unique_keypoints(
                 i=i2, keypoints=keypoints_i2, per_image_kpt_coordinates=per_image_kpt_coordinates
             )
-            putative_corr_idxs = np.stack([i1_indices, i2_indices], axis=-1).astype(np.uint16)
+            putative_corr_idxs = np.stack([i1_indices, i2_indices], axis=-1).astype(np.int32)
```
Comment (travisdriver) [inline gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py:108]: Why this change? indices should be unsigned

**Review by travisdriver — APPROVED** [review]
Please fix the type hint before merging.

File: gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py (hunk 6 above)
Comment (johnwlambert) [reply gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py:108]: the gtsam MatchIndicesMap for C++ DSF currently requires int32 input for Eigen (https://github.com/borglab/gtsam/blob/develop/gtsam/sfm/DsfTrackGenerator.h#L32)

File: gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py (hunk 5 above)
Comment (johnwlambert) [reply gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py]: Fixed, thanks.
